pandas/tests/reshape/test_concat.py

```diff
@@ -2776,3 +2776,137 @@ def test_concat_datetimeindex_freq():
     expected = pd.DataFrame(data[50:] + data[:50], index=dr[50:].append(dr[:50]))
     expected.index.freq = None
     tm.assert_frame_equal(result, expected)
+
+
+@pytest.mark.parametrize("suffixes", ["_a", ("_x"), ["a", "b"]])
+def test_concat_suffixes_type(suffixes):
+    # GH 21791, like pd.merge, here suffixes type should be tuple
+    objs = [pd.Series([1, 2], name="a"), pd.DataFrame({"a": [2, 3]})]
+    with pytest.raises(ValueError, match="only 'tuple' is allowed"):
+        pd.concat(objs, axis=1, suffixes=suffixes)
+
+
+@pytest.mark.parametrize(
+    "objs, suffixes",
+    [
+        (
+            [
+                pd.Series([1, 2], name="a"),
+                pd.Series([2, 3], name="a"),
+                pd.Series([2, 3]),
+            ],
+            ("_x", "_y"),
+        ),
+        (
+            [
+                pd.DataFrame({"a": [1, 2]}),
+                pd.DataFrame({"a": [2, 3]}, pd.Series([1, 2])),
+            ],
+            ("_x", "_y", "_z", "_k"),
+        ),
+        (
+            [pd.DataFrame({"a": [1, 2]}), pd.DataFrame({"a": [2, 3]})],
+            ("_x", "_y", "_z"),
+        ),
+    ],
+)
+def test_concat_suffixes_length_unmatch_error(objs, suffixes):
+    # GH 21791, add test to see if warning is raise when columns overlap but length of
+    # suffixes does not match the length of objs
+    with pytest.raises(ValueError, match="Number of objects for concatenation is not"):
+        pd.concat(objs, axis=1, suffixes=suffixes)
+
+
+@pytest.mark.parametrize(
+    "objs, suffixes, expected",
+    [
+        (
+            [pd.Series([1, 2], name="a"), pd.Series([2, 3], name="a")],
+            ("_x", "_y"),
```

Comment:
Can you also add tests for if dupe `suffixes` are specified, e.g. `("_x", "_x")`?